runtime/include/spinlock.hpp

```diff
@@ -0,0 +1,59 @@
+#ifndef SPINLOCK_HPP
+#define SPINLOCK_HPP
+
+#include <atomic>
+#include <cstddef>
+#include <ctime>
+#include <time.h>
+
+class Spinlock {
+    std::atomic<unsigned int> lock;
+public:
+
+    // Default constructor to initialize the atomic_flag
+    Spinlock() : lock(0) {}
+
+    // Deleted copy constructor to prevent copying
+    Spinlock(const Spinlock&) = delete;
+    Spinlock& operator=(const Spinlock&) = delete;
+
+    Spinlock(Spinlock&& other) noexcept : lock() {
+        if (other.lock.exchange(1, std::memory_order_acquire)) {
+            lock.store(0, std::memory_order_release);
+        }
+    }
+
+    Spinlock& operator=(Spinlock&& other) noexcept {
+        if (this != &other) {
+            if (other.lock.exchange(1, std::memory_order_acquire)) {
+                lock.store(0, std::memory_order_release);
+            }
+        }
+        return *this;
+    }
+
+    void spin_lock() {
+        static const timespec ns = {0, 1};
+        for(int i = 0; lock.load(std::memory_order_relaxed) || lock.exchange(1, std::memory_order_acquire); ++i) {
```

Comment:
Variable `i` will overflow. Overflow of signed integers are undefined behaviors. Undefined behaviors may lead to incorrect compiler optimizations

https://stackoverflow.com/questions/16188263/is-signed-integer-overflow-still-undefined-behavior-in-c

Reply:
will it overflow though? there is a check for `i==8`?

Reply:
My fault. Sorry